Composer/packages/form-dialogs/src/undo/UndoRoot.tsx

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import {
+  Loadable,
+  RecoilState,
+  Snapshot,
+  useGotoRecoilSnapshot,
+  useRecoilSnapshot,
+  // eslint-disable-next-line @typescript-eslint/camelcase
+  useRecoilTransactionObserver_UNSTABLE,
+} from 'recoil';
+import * as React from 'react';
+import { AtomMap, History } from 'src/undo/types';
+
+const getAtomMap = (snapshot: Snapshot, trackedAtoms: RecoilState<any>[]): AtomMap => {
+  const atomMap = new Map<RecoilState<any>, Loadable<any>>();
+  for (const atom of trackedAtoms) {
+    atomMap.set(atom, snapshot.getLoadable(atom));
+  }
+  return atomMap;
+};
+
+const didAtomMapsChange = (prevMap: AtomMap, currMap: AtomMap) => {
+  if (prevMap.size !== currMap.size) {
+    return true;
+  }
+
+  for (const key of prevMap.keys()) {
+    if (!currMap.has(key)) {
+      return true;
+    }
+
+    const prevVal = prevMap.get(key);
+    const currVal = currMap.get(key);
+
+    if (prevVal && currVal) {
+      if (prevVal.state !== currVal.state) {
+        return true;
+      }
+
+      if (prevVal.state === 'hasValue' && currVal.state === 'hasValue' && prevVal.contents !== currVal.contents) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+};
+
+const mapTrackedAtomsOntoSnapshot = (
+  current: Snapshot,
+  target: Snapshot,
+  trackedAtoms: RecoilState<any>[] | null | undefined
+): Snapshot => {
+  if (!trackedAtoms) {
+    return target;
+  }
+
+  const atomMap = getAtomMap(target, trackedAtoms);
+
+  return current.map((pendingSnap) => {
+    for (const [atom, loadable] of atomMap.entries()) {
+      if (loadable.state === 'hasValue') {
+        pendingSnap.set(atom, loadable.contents);
+      }
+    }
+  });
+};
+
+type ContextState = {
```

Comment:
Do you want to export the ContextState type so that callers can useContext with it?